velox/exec/HashAggregation.cpp

```diff
@@ -186,12 +186,13 @@ void HashAggregation::prepareOutput(vector_size_t size) {
 }
 
 void HashAggregation::flushPartialOutputIfNeed() {
-  if (partialFull_) {
-    stats().addRuntimeStat(
-        "flushRowCount", RuntimeCounter(groupingSet_->numRows()));
-    groupingSet_->resetPartial();
-    partialFull_ = false;
+  if (!partialFull_) {
+    return;
   }
+  stats().addRuntimeStat("flushRowCount", RuntimeCounter(numOutputRows_));
```

Comment:
It is not obvious to me why `groupingSet_->numRows()` is different from numOutputRows_. Any chance you could explain a bit more?

It is also not clear what's the benefit of flushing partial global aggregation. If we run out of member in partial global aggregation, won't we run out of memory in final global aggregation as well? Spilling won't help because there is only one row.

Reply:
'It is not obvious to me why groupingSet_->numRows() is different from numOutputRows_. Any chance you could explain a bit more?'
There are two problems in using groupingSet_->numRows(): (1) in case of global aggregation, it always return zero as table is always null; (2) in case of group by, the table rows have always be cleared by GroupingSet::getOutput right before we call flushPartialOutputIfNeed. The table clear code is here: https://github.com/facebookincubator/velox/blob/main/velox/exec/GroupingSet.cpp#L412 (the returned numGroups will be zero) and verified in unit test: https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/AggregationTest.cpp#L790

' what's the benefit of flushing partial global aggregation'
The existing code does partial global aggregation flush. This PR just fixes the global aggregation state after flush. That's a good point. I guess we might expect the final aggregation have a larger buffer size or auto extend the buffer size for final aggregation at least?

Reply:
> I guess we might expect the final aggregation have a larger buffer size or auto extend the buffer size for final aggregation at least?

I don't believe that's the case. It seems that we should not allow flushing partial state for global aggregations.

Reply:
>  in case of group by, the table rows have always be cleared by GroupingSet::getOutput right before we call flushPartialOutputIfNeed

It sounds like 'flushPartialOutputIfNeed' is named incorrectly. If there is not output already, then what would flush-output do? Would it make sense to rename this method for clarity?

Reply:
Yeah, how about we change it to resetPartialOutputIfNeed? Thanks!

Reply:
> > I guess we might expect the final aggregation have a larger buffer size or auto extend the buffer size for final aggregation at least?
> 
> I don't believe that's the case. It seems that we should not allow flushing partial state for global aggregations.

So given that, we just use as much memory as possible for global aggregation and we might have the risk to run into OOM since we don't do disk spilling for global partial aggregation? I mean at the final aggregation stage, we can rely on disk spilling to avoid that? Thanks!

Reply:
That would be a better name.

Reply:
> > I guess we might expect the final aggregation have a larger buffer size or auto extend the buffer size for final aggregation at least?
> 
> I don't believe that's the case. It seems that we should not allow flushing partial state for global aggregations.

So basically, in global aggregation case, we have no choice other than bump up the buffer size at both partial and final stage? Thanks!

Reply:
> at the final aggregation stage, we can rely on disk spilling to avoid that?

I don't believe it is possible to do spilling for global aggregation, since there is only one row and spilling is row-based.

Reply:
> > at the final aggregation stage, we can rely on disk spilling to avoid that?
> 
> I don't believe it is possible to do spilling for global aggregation, since there is only one row and spilling is row-based.

Yeah, it makes sense. https://github.com/facebookincubator/velox/pull/2106 disable the flush for global aggregation.